### PR TITLE
`lpinfo -m` timeout parallelization of PPD Generators.

### PR DESCRIPTION
**Pull request by pranjanpr**
This PR is with respect to issue Issue https://github.com/OpenPrinting/cups/issues/65 .

`driverless` was taking too much time, so created a separate thread to call the PPD generators while waiting for timeout in the main thread. If the PPD is not returned within the timeout then terminated that call and looped over.
I have created separate functions regarding this. Requires some testing.

Commit 1204a8b
@michaelrsweet @tillkamppeter Please review this.

### CODE REVIEW

File: cups/dest.c [hunk 1]
```diff
@@ -4183,14 +4191,14 @@ cups_get_dests(
       if ((dest = cupsGetDest(name, instance, num_dests, *dests)) == NULL)
       {
        /*
-	* Out of memory!
-	*/
+	      * Out of memory!
+	      */
```
Comment (zdohnal) [inline cups/dest.c:4195]: @pranjanpr looks like unintended change, would you mind reverting it?

You can always check the current changes in git before commit them by ```git diff``` (if the changes haven't been staged yet) or ```git diff --staged``` (if you already staged the changes).

File: cups/dest.c [hunk 2]
```diff
@@ -4183,14 +4191,14 @@ cups_get_dests(
...
 
         DEBUG_puts("9cups_get_dests: Out of memory!");
         break;
       }
     }
-
+
```
Comment (zdohnal) [inline cups/dest.c:4201]: @pranjanpr the same as previous.

File: scheduler/cups-driverd.cxx [hunk 3]
```diff
@@ -113,11 +115,17 @@ typedef union				/**** TAR record format ****/
   }	header;
 } tar_rec_t;
 
+struct args {
+    cups_file_t *fp;
+    char *buf;
+    size_t buflen;
+};
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:122]: In CUPS we use ```typedef``` for defining structures, see [coding guidelines](https://github.com/OpenPrinting/cups/blob/master/DEVELOPING.md) regarding types.
Other thing is a formatting - CUPS uses spaces till 7 characters, but tab(s) must be used instead of 8 spaces, 16 spaces etc. (so f.e. if you do the indentation for 10 spaces, you will use 1 tab and 2 spaces...).

File: cups/dest.c [hunk 4]
```diff
@@ -3502,6 +3502,12 @@ cups_enum_dests(
 
     num_dests = _cupsGetDests(http, IPP_OP_CUPS_GET_PRINTERS, NULL, &dests, type, mask);
 
+    /*
+    * Pass the number of destinations to fetch required dest...
+    */
+
+    num_dests = cups_get_dests(filename,NULL, NULL, 1, user_default != NULL, num_dests, &dests);
```
Comment (zdohnal) [inline cups/dest.c:3509]: @pranjanpr there must be a space after comma in arguments.

File: cups/dest.c [hunk 5]
```diff
@@ -3688,6 +3694,8 @@ cups_enum_dests(
   else
     remaining = msec;
 
+  data.num_dests = cups_get_dests(filename ,NULL, NULL, 1, user_default != NULL, data.num_dests, &data.dests);
```
Comment (zdohnal) [inline cups/dest.c:3697]: @pranjanpr there mustn't be a space between an argument and a comma, the space must be after the comma.

File: scheduler/cups-driverd.cxx [hunk 6]
```diff
@@ -113,11 +115,17 @@ typedef union				/**** TAR record format ****/
...
 
 /*
  * Globals...
  */
 
+static char* retval=NULL;
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:128]: @pranjanpr Global variables must be capitalized and have a comment what they are meant for. And IMO ```retval``` is too generic name.

File: scheduler/cups-driverd.cxx [hunk 7]
```diff
@@ -183,6 +191,11 @@ static int		read_tar(cups_file_t *fp, char *name, size_t namesize,
 			         struct stat *info);
 static regex_t		*regex_device_id(const char *device_id);
 static regex_t		*regex_string(const char *s);
+void              *cupsFileGetsCall(void *input);
+char              *cupsFileGetsT(cups_file_t *fp, char*buf, size_t buflen);
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:195]: @pranjanpr indentation looks off, would you mind looking into it!

File: scheduler/cups-driverd.cxx [hunk 8]
```diff
@@ -183,6 +191,11 @@ static int		read_tar(cups_file_t *fp, char *name, size_t namesize,
...
+
+
+
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:198]: @pranjanpr those new lines are unnecessary.

File: scheduler/cups-driverd.cxx [hunk 9]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
 
   return (NULL);
 }
+
+void *
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2931]: @pranjanpr IIUC you only return NULL and I'm not sure why it is needed. You only check ```retval``` later, but never the return value of ```cupsFileGetsCall()```).
IMO you can drop ```void *``` and ```return NULL;``` and use only ```void``` instead.

File: scheduler/cups-driverd.cxx [hunk 10]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+cupsFileGetsCall(void *input)
+{
+  retval = cupsFileGets(((struct args*)input)->fp,((struct args*)input)->buf,((struct args*)input)->buflen);
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2934]: @pranjanpr you can do a following:

1) if you defined your struct with typedef before, you can use 'some_type_t' instead of writing ```struct args```
2) you can do a recast before calling ```cupsFileGets()``` into a local variable and then use a local variable for arguments. IMO it will look better than three recasts in arguments.

File: scheduler/cups-driverd.cxx [hunk 11]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+  return NULL;
+}
+
+/*
+  * Alternate to cupsFileGets() 
+  */
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2940]: @pranjanpr would you mind expanding the comment? Please look into other comments for CUPS functions.
Indentation doesn't match, would you mind looking into it?

File: scheduler/cups-driverd.cxx [hunk 12]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+
+char * 
+cupsFileGetsT(cups_file_t *fp,		/* I - CUPS file */
+             char        *buf,		/* O - String buffer */
+	     size_t      buflen)        /* I - Size of string buffer */
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2945]: @pranjanpr Indentation doesn't match, would you mind looking into it?

File: scheduler/cups-driverd.cxx [hunk 13]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+  struct timeval t1, t2;
+
+  struct args *arg = (struct args *)malloc(sizeof(struct args));
+  arg->fp = fp;
+  arg->buf = buf;
+  arg->buflen = buflen;
+
+  gettimeofday(&t1,NULL);
+
+  pthread_create(&ptid,NULL,cupsFileGetsCall,(void*)arg);
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2964]: @pranjanpr  missing spaces between commas and next arguments.

File: scheduler/cups-driverd.cxx [hunk 14]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+
+  pthread_t ptid;
+  struct timeval t1, t2;
+
+  struct args *arg = (struct args *)malloc(sizeof(struct args));
+  arg->fp = fp;
+  arg->buf = buf;
+  arg->buflen = buflen;
+
+  gettimeofday(&t1,NULL);
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2962]: @pranjanpr missing space after comma.

File: scheduler/cups-driverd.cxx [hunk 15]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+  int flag = 0;
+
+ /*
+  * Creating a new thread ID.
+  */
+
+  pthread_t ptid;
+  struct timeval t1, t2;
+
+  struct args *arg = (struct args *)malloc(sizeof(struct args));
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2957]: @pranjanpr if you define the struct via typedef, you can use the new type you created.

File: scheduler/cups-driverd.cxx [hunk 16]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+  pthread_create(&ptid,NULL,cupsFileGetsCall,(void*)arg);
+
+ /*
+  * Loop through cupsFileGets() until timeout (2 sec)
+  */
+
+  while(retval == NULL)
+  {
+    gettimeofday(&t2, NULL);
+    double elapsedTime = (t2.tv_sec - t1.tv_sec);
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2973]: @pranjanpr IIUC no need for brackets.

File: scheduler/cups-driverd.cxx [hunk 17]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+
+    if(elapsedTime > 2)
+    {
+      flag=1;
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:2977]: @pranjanpr missing spaces between ```=```.

File: scheduler/cups-driverd.cxx [hunk 18]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
+    fprintf(stderr,"%s\n","Thread terminated");
+    pthread_cancel(ptid);
+
+    retval = NULL;
+  }
+  else
+    pthread_join(ptid, NULL);
+
+return retval;
+}
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:3000]: @pranjanpr missing newline on the end of the file - it's common to end a file with new line.

File: scheduler/cups-driverd.cxx [hunk 19]
```diff
@@ -22,7 +22,9 @@
 #include <cups/ppd-private.h>
 #include <ppdc/ppdc.h>
 #include <regex.h>
-
+#include <pthread.h>
+#include <sys/time.h>
+#include <unistd.h>
```
Comment (zdohnal) [inline scheduler/cups-driverd.cxx:27]: @pranjanpr please sort all includes alphabetically once you add a new. First are local project includes (uses double quotes), then system includes (uses <>).

**Comment by zdohnal** [comment]
I did a review regarding code style, @pranjanpr please check the comments. Most comments are based on [CUPS developer guidelines](https://github.com/OpenPrinting/cups/blob/master/DEVELOPING.md).

**Comment by pranjanpr** [comment]
@zdohnal 
I will revert back as soon as I correct all the corrections. Thanks for reviewing it thoroughly.

File: cups/dest.c [hunk 20]
```diff
@@ -3502,6 +3502,12 @@ cups_enum_dests(
 
     num_dests = _cupsGetDests(http, IPP_OP_CUPS_GET_PRINTERS, NULL, &dests, type, mask);
 
+    /*
```
Comment (michaelrsweet) [inline cups/dest.c:3505]: These changes seem unrelated to making cups-driverd run the driver interface programs in parallel. Please separate your changes (probably each on their own branch in your repository).

File: scheduler/cups-driverd.cxx [hunk 21]
```diff
@@ -113,11 +115,17 @@ typedef union				/**** TAR record format ****/
   }	header;
 } tar_rec_t;
 
+struct args {
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:118]: Like @zdohnal says, see the coding guidelines for the naming conventions and coding style we use

Wrap the struct in a typedef.  "args" isn't a meaningful name for a structure, and there are no comments to tell me what the structure or members are used for.

File: scheduler/cups-driverd.cxx [hunk 22]
```diff
@@ -183,6 +191,11 @@ static int		read_tar(cups_file_t *fp, char *name, size_t namesize,
 			         struct stat *info);
 static regex_t		*regex_device_id(const char *device_id);
 static regex_t		*regex_string(const char *s);
+void              *cupsFileGetsCall(void *input);
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:194]: I'm not sure why these functions have a "cupsFile" prefix - they aren't part of libcups or the cups_file_t API.

File: scheduler/cups-driverd.cxx [hunk 23]
```diff
@@ -1842,7 +1855,7 @@ load_drivers(cups_array_t *include,	/* I - Drivers to include */
 
     if ((fp = cupsdPipeCommand(&pid, filename, argv, 0)) != NULL)
     {
-      while (cupsFileGets(fp, line, sizeof(line)))
+      while (cupsFileGetsT(fp, line, sizeof(line)))
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:1858]: What does "cupsFileGetsT" do?

File: scheduler/cups-driverd.cxx [hunk 24]
```diff
@@ -2914,3 +2927,74 @@ regex_string(const char *s)		/* I - String to compare */
...
 }
+
+void *
+cupsFileGetsCall(void *input)
+{
+  retval = cupsFileGets(((struct args*)input)->fp,((struct args*)input)->buf,((struct args*)input)->buflen);
+  return NULL;
+}
+
+/*
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:2938]: OK, so cupsFileGetsT implements a timed out read using a thread? This is totally unnecessary as UNIX-based systems can use `select` or `poll` (easier) to wait for input on the file descriptor (retrieved using cupsFileNumber).

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
@pranjanpr So it isn't clear how these changes actually run the driver interface programs in parallel. I would expect to see a loop that creates a reader thread after each call to cupsdPipeCommand and then some code that manages an overall maximum number of seconds of run time.

**Comment by pranjanpr** [comment]
> @pranjanpr So it isn't clear how these changes actually run the driver interface programs in parallel. I would expect to see a loop that creates a reader thread after each call to cupsdPipeCommand and then some code that manages an overall maximum number of seconds of run time.

@michaelrsweet Sir, I actually did not created different threads to divide the driver files among them, instead the cupsFileGetsT() function creates a main thread with a timeout and an another thread to get the PPD files using cupsFileGets() function. If the thread is unable to return within the timeout range then the main thread cancels the separate thread.
 I will work on the requested changes and revert soon with all the corrections.

**Comment by pranjanpr** [comment]
This PR is closed because my origin fork was from `apple/cups` and `OpenPrinting/cups` is no longer following it. All the requested changes have been corrected. Please consider this https://github.com/OpenPrinting/cups/pull/114
